### ISSUE TITLE
SMTP authentication tracebacks when receiving an empty line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2521]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#2521|
|Type|defect|
|Created|2007-03-13 21:02:31Z|

state_AUTH takes the empty string to have a special meaning.

<details><summary>Searchable metadata</summary>

```
trac-id__2521 2521
type__defect defect
reporter__washort washort
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1173819751000000 1173819751000000
changetime__1174404305000000 1174404305000000
version__None None
owner__washort washort

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_1"></a>|@washort commented|
|-|-|

Changed it to use None instead. Please review in empty-password-2521.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Stealing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @washort<br>@glyph set status to `new`|
|-|-|


  * The test really ought to verify that you can actually _log in_ with an empty password, rather than just verifying that no exception happens.  That is a valid password, after all.
  * The parameter specification for ```state_AUTH``` changed.  That means its docstring needs updating.  Explain the meaning of string vs. '*' vs. None, etc.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_4"></a>|@washort set owner to @glyph|
|-|-|

Fixed. Please re-review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @washort|
|-|-|

Trailing whitespace follows ```test_plainAuthenticationEmptyPasswd```, and there should be 2 blank lines there, not 1, as per the coding standard.

Also, please make it "password", not "passwd" in the test name.  That abbreviation isn't used anywhere else in the test module, or the implementation.

```test_plainAuthenticationInitialResponse``` uses ```assertServerAuthenticated```.  I think this test should do that as well.  In fact, it might make sense to refactor that test so that the username and password could be passed to assertServerAuthenticated, so that the body of the test didn't need to be repeated.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_6"></a>|@washort set owner to @glyph|
|-|-|

Fixed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @washort|
|-|-|

OK.  Looking good.  Merge.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort set status to `closed`|
|-|-|

(In [19881]) Merge empty-password-2521.
Author: washort
Reviewer: glyph
Fixes [#2521](https://github.com/twisted/twisted/issues/2521).

Handle empty passwords in SMTP auth properly.
